### ISSUE TITLE
a couple of suggestions for v5

### ISSUE THREAD

**Issue by novaknole** [OpenZeppelin/openzeppelin-contracts#3870]
**🧐 Motivation**

I have a couple of questions/suggestions and as pointed out from the forum where I initially posted them, I also post them here.

**Number 1:**
Currently, `upgradeToAndCall` seems to me `external virtual`. This causes a problem where I can't override it and call it inside my function.. E.x

```
function upgradeToAndCall(...) external override {
    // my custom logic
   super.upgradeToAndCall(); 
}
```

In the above example, `super.upgradeToAndCall(); ` becomes not possible because of `external` visibility. I'd suggest making it public.

**Number 2:** 

If implementation passed to upgradeTo is the same as the currently set one, why do you still emit the event ? I understand that It could be the caller's responbility to not call `upgradeTo` if the addresses match, but I'd suggest not emitting the event if that still happens. Not a big deal, just an idea.

**Number 3:**

Hopefuly, in `v5` version, we will see `reinitializer` to accept `uint256` instead of uint8. I understand why you did uint8 to allow compatibility and not breaking storage gap, but for v5, making it `uint256` is the right way to go.

**Comment by frangio** [comment]
Thanks @novaknole!

1. @Amxx Did we have a reason for making `upgradeToAndCall` external and not public? This is not a breaking change so we could change it before 5.0.
2. This one might be controversial. Generally I would emit events if they don't represent a change. For example, an ERC20 `Transfer` event should be emitted even if the value is 0. There are places where we don't follow this guideline though, such as AccessControl's grant and revoke functions (we could revise those). In the case of the upgrade functions, I also think that the events should be emitted always, in particular because of the "and call" part, so that a reinitialization would become evident in the events.
3. Interesting... We hadn't considered this yet. It still makes sense to pack both "initializing" and "initialized version" values in one slot, so we may do `uint248`.

These suggestions should be made into individual issues if we agree with them.

**Comment by novaknole** [comment]
Hey @frangio 

3. The reason I don't like `uint8` is that I can't integrate downgrading mechanism for plugins. For example, I wanted to use semantic versioning for plugin versions. Let's say I pushed 1.3.0 to the registry.  My framework would transform this 1.3.0 into the single integer automatically so users don't have to worry about what to put in reinitializer. Then I push 1.5.0, then I push 1.4.0. Problem is because of `uint8`, The possibilities of semantic versioning are not much. The last time I calculated, I could do max 4.7.7 due to 255 possibilities and I couldn't push 5.1.0 for example. `uint256` solves everything. `uint248` would be good as well.
